packages/flutter/lib/src/services/text_editing_delta.dart

```diff
@@ -0,0 +1,424 @@
+      return TextEditingDeltaInsertion(
+        oldText: oldText,
+        textInserted: replacementSource.substring(replacementDestinationEnd - replacementDestinationStart, (replacementDestinationEnd - replacementDestinationStart) + (replacementSource.length - (replacementDestinationEnd - replacementDestinationStart))),
+        insertionOffset: replacementDestinationEnd,
+        selection: newSelection,
+        composing: newComposing,
+      );
+    } else if (isReplaced) {  // Replacement.
+      return TextEditingDeltaReplacement(
+        oldText: oldText,
+        replacementText: replacementSource,
+        replacedRange: TextRange(
+          start: replacementDestinationStart,
+          end: replacementDestinationEnd,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    }
+    assert(false);
+    return TextEditingDeltaNonTextUpdate(
+      oldText: oldText,
+      selection: newSelection,
+      composing: newComposing,
+    );
+  }
+
+  /// The old text state before the delta has occurred.
+  final String oldText;
+
+  /// The range of text that is currently selected after the delta has been
+  /// applied.
+  final TextSelection selection;
+
+  /// The range of text that is still being composed after the delta has been
+  /// applied.
+  final TextRange composing;
+
+  /// {@template flutter.services.TextEditingDelta.apply}
+  /// This method will take the given [TextEditingValue] and return a new
+  /// [TextEditingValue] with that instance of [TextEditingDelta] applied to it.
+  /// {@endtemplate}
+  TextEditingValue apply(TextEditingValue value);
+}
+
+/// The delta is inserting a single/or contigous sequence of characters.
+@immutable
+class TextEditingDeltaInsertion extends TextEditingDelta {
+  /// Creates an insertion delta for a given change to the editing state.
+  ///
+  /// {@macro flutter.services.TextEditingDelta}
+  const TextEditingDeltaInsertion({
+    required String oldText,
+    required this.textInserted,
+    required this.insertionOffset,
+    required TextSelection selection,
+    required TextRange composing,
+  }) : super(
+      oldText: oldText,
+      selection: selection,
+      composing: composing,
+  );
+
+  /// The text that is being inserted into [oldText].
+  final String textInserted;
+
+  /// The offset in the [oldText] where the insertion begins.
+  final int insertionOffset;
+
+  /// {@macro flutter.services.TextEditingDelta.apply}
+  @override
+  TextEditingValue apply(TextEditingValue value) {
+    // To stay inline with the plain text model we should follow a last write wins
+    // policy and apply the delta to the oldText. This is due to the asyncronous
+    // nature of the connection between the framework and platform text input plugins.
+    try {
+      if (oldText != value.text) {
+        throw FlutterError('The editing state is out of sync between the framework and the engine.');
```

Comment:
I think you can just call `FlutterError.reportError` here? Also the `oldText != value.text` should be in an assert probably.

Reply:
I'm still not completely convinced that this should be an error. Conflicts can happen (especially when the developer has additional input formatters/mutating onChanged callbacks), and the error will probably get spammed in these scenarios. Also it is typically not a programmer error thus not actionable (and a little bit confusing).

Reply:
I'm ok with just a comment here then.  We can revisit this if it ever seems to be a problem for developers.

Reply:
I agree that this may lead to a lot of log spam. I though this log would be useful, but if the developer cannot act on it there doesn't seem to be a good reason to have it. Maybe just the comment will suffice? 

If a developer wanted to act on this asynchronous failure then they would have to extend their desired `TextEditingDelta` implementation, and capture the error themselves in the apply method and have logic to handle it. For example say they wanted to see if this happens with an insertion. They would extend TextEditingDeltaInsertion and override apply. They would convert any `TextEditingDeltaInsertion` to their new `TextEditingDeltaNewInsertion` and apply those to their editing value instead of the original ones.